Add contribution docs on adding component configuration options (#3387)
alphagov/govuk-frontend: docs/contributing/component-js-configuration-options.md

# Adding JavaScript configuration options to components

> Configuration options permit developers to pass additional information into a component's JavaScript, allowing them to change how the component might work or appear without needing to create their own fork.

Reviewer: ```suggestion
Configuration options allow developers to pass additional information into a component's JavaScript. This allows them to change how the component might work or appear without needing to create their own fork.
```

> Configuration options straddle the line between Nunjucks, HTML and JavaScript.

Reviewer: What does 'straddle the line' mean in this context? Is it the Nunjucks, HTML and JavaScript all use the same configuration options or that they can recognise each others configuration options or something similar?

Author: I guess what I meant here is that—although configuration options are only currently used for JavaScript—the manner in which they are set exists across all three of these contexts.

I was trying to set up the next few sections, basically 😅

Reviewer: ok. So could we use your sentence something like this?
Currently, only JavaScript uses the configuration options, but due to the way they are written Nunjucks and HTML can use them as well.
Is that accurate?

Author: It's more like the options can be configured using Nunjucks, HTML or JavaScript, but only JavaScript will use them?

Reviewer: Oh I see. I had it almost totally backwards! Yes, that sounds good.

> Configuration options straddle the line between Nunjucks, HTML and JavaScript.

Reviewer: ```suggestion
Configuration options can be configured using Nunjucks, HTML or JavaScript, but only JavaScript will use the options.
```

## Preparing the component's JavaScript

> First, ensure the entry function for component has a parameter for passing in a configuration object.

Reviewer: ```suggestion
First, make sure the entry function for the component has a parameter for passing in a configuration object.
```

```javascript
function Accordion ($module, config) {
  ...
}
```

Within the entry function, you'll then create an object with the default configuration for the component. These are what options will be used if no other options have been provided.

```javascript
var defaultConfig = {
  rememberExpanded: true
}
```

> Next, use the `mergeConfigs` helper to combine the default config and the configuration object. This is assigned globally (using `this`) so that it's accessible anywhere within the component's JavaScript.

Reviewer: Would it work to structure this doc with numbered steps in each section?

Reviewer: Where it says 'This is assigned globally...' is it referring to the helper, the default config, the configuration object or the result of the merging done by the helper?

Author: > Would it work to structure this doc with numbered steps in each section?

I considered this, but as some steps contain multiple paragraphs of information and code blocks, it isn't easy to semantically mark it up as a numbered list, so I didn't. 

> Where it says 'This is assigned globally...' is it referring to the helper, the default config, the configuration object or the result of the merging done by the helper?

The result of the merging done by the helper.

Reviewer: Yeah, as  worked my way down I could see that numbering would only work in some places making the whole doc rather inconsistent.

> The result of the merging done by the helper.

Thanks

> Next, use the `mergeConfigs` helper to combine the default config and the configuration object. This is assigned globally (using `this`) so that it's accessible anywhere within the component's JavaScript.

Reviewer: ```suggestion
Next, use the `mergeConfigs` helper to combine the default config and the configuration object. The result of the merge is assigned globally (using `this`) so that it's accessible anywhere within the component's JavaScript.
```

> The order that variables are written defines their priority, with objects passed sooner being overwritten by those passed later. As we want the user's configuration to take precedence over our defaults, we list our default configuration object first.

Reviewer: ```suggestion
The order in which variables are written defines their priority, with objects passed sooner being overwritten by those passed later. As we want the user's configuration to take precedence over our defaults, we list our default configuration object first.
```

> We also do a little safety check here. There is no guarantee that `config` will have any value at all, in which case it'll be `undefined`. We use an OR operator (`||`) to quickly check if this is the case and, if so, we use an empty object instead.

Reviewer: ```suggestion
There is no guarantee `config` will have any value at all, so it'll be `undefined`. We use an OR operator (`||`) as a safety check. If the value is `undefined`, we use an empty object instead.
```

```javascript
import { mergeConfigs } from '../../common/index.mjs'

this.config = mergeConfigs(
  defaultConfig,
  config || {}
)
```

We can now reference the configuration option anywhere we need to in the component's JavaScript:

```javascript
this.config.rememberExpanded
// => true
```

> It's now possible to individually initialising the component with configuration options.

Reviewer: ```suggestion
It's now possible to individually initialise the component with configuration options.
```

```html
<script>
  var $accordion = document.getElementById("accordion")
  new window.GOVUKFrontend.Accordion($accordion, {
    rememberExpanded: false
  }).init()
</script>
```

## Allowing options to be passed through the `initAll` function

Usually, teams will not be individually initialising components. Instead, GOV.UK Frontend ships with an `initAll` function, which searches the page for instances of components and automatically initialises them.

In `src/govuk/all.mjs`, update the component's `new` function to pass through a nested configuration object. The nested object should use the component's name converted to camelCase (for example, the 'Character Count' component becomes `characterCount`).

```javascript
new Accordion($accordion, config.accordion).init()
```

> It's now possible to pass configuration options for your component, as well as multiple other components, via the `initAll` function.

Reviewer: ```suggestion
It's now possible to pass configuration options for your component, as well as multiple other components, using the `initAll` function.
```

```html
<script>
  window.GOVUKFrontend.initAll({
    accordion: {
      rememberExpanded: false
    },
    characterCount: {
      threshold: 90
    }
  })
</script>
```

> ## Adding support for HTML `data-*` attributes

Reviewer 2: Mind if we link to this section from the JS coding standards?
https://github.com/alphagov/govuk-frontend/blob/main/docs/contributing/coding-standards/js.md#use-data-attributes-to-initialise-component-javascript

It's not been updated beyond `data-module`

> For convenience, we allow for configuration options to be passed through HTML `data-*` attributes too.

Reviewer: ```suggestion
For convenience, we also allow for configuration options to be passed through HTML `data-*` attributes.
```

> `data-*` attributes can be found in JavaScript by looking at an element's `dataset` property. Browsers will helpfully convert the attribute names from HTML's kebab-case to more JavaScript-friendly camelCase when working with `dataset`. See ['Naming configuration options'](#naming-configuration-options) for caveats.

Reviewer: ```suggestion
You can find `data-*` attributes in JavaScript by looking at an element's `dataset` property. Browsers will convert the attribute names from HTML's kebab-case to more JavaScript-friendly camelCase when working with `dataset`. See ['Naming configuration options'](#naming-configuration-options) for exceptions.
```

> As we expect configuration-related `data-*` attributes to always be on the component's root element (the same element with the `data-module` attribute), we can access them all quickly using `$module.dataset`.

Reviewer: ```suggestion
As we expect configuration-related `data-*` attributes to always be on the component's root element (the same element with the `data-module` attribute), we can access them all using `$module.dataset`.
```

> Let's update our `mergeConfigs` call from earlier to include `$module.dataset` as the highest priority.

Reviewer: ```suggestion
Using the `mergeConfigs` call discussed earlier in this document, update it to include `$module.dataset` as the highest priority.
```

```javascript
import { mergeConfigs } from '../../common/index.mjs'
import { normaliseDataset } from '../../common/normalise-dataset.mjs'

> this.config = mergeConfigs(

Reviewer 2: Until we can use `Object.assign()` (or object spread) we should suggest types here 😩 

For example in the Character Count component, linting errors appear for:

```js
this.config.maxlength
this.config.threshold
```

Because without types, we appear to do calculations on unknown types

```suggestion
/** @type {CharacterCountConfig} */
this.config = mergeConfigs(
```

Reviewer 2: Hard isn't it

Also the realisation that `normaliseDataset()` is a bit magic and will automatically picks types for you based on what the string value looks like. The compiler can't automatically "type" the return value.

Could potentially be solved by passing the defaults into `normaliseDataset()`

```js
this.config = normaliseDataset($module.dataset, {
  ...defaultConfig,
  ...config || {},
})
```

So rather than guessing types you could grab them from defaults

Not one for this PR, so yeah let's leave it

  defaultConfig,
  config || {},
  normaliseDataset($module.dataset)
)
```

> Here, we pass it through our `normaliseDataset` function. This is because attribute values in dataset are always interpreted as being strings. `normaliseDataset` runs a few simple checks to convert values back to numbers or booleans where appropriate.

Reviewer: ```suggestion
Here, we pass `mergeConfigs` through our `normaliseDataset` function. This is because attribute values in dataset are always interpreted as strings. `normaliseDataset` runs a few simple checks to convert values back to numbers or booleans where appropriate.
```

Author: This isn't quite accurate. We're passing the value of `$module.dataset` through the `normaliseDataset` function, not the value or result of `mergeConfigs`. 

The code example does this inside the `mergeConfigs` function essentially for brevity, but it doesn't strictly have to be nested in this way.

Reviewer: ok, so if we change 'it' to 'the value of `$module.dataset`', do we need to add further explanation or is that sufficient by itself?

Author: I think that should be sufficient, yeah

Reviewer: Fab, thank you

Now, in our HTML, we could pass configuration options by using the kebab-case version of the option's name.

```html
<div
  class="govuk-accordion"
  data-module="govuk-accordion"
  data-remember-expanded="false">
  ...
</div>
```

> However, this only works for developers who are writing raw HTML. We include Nunjucks macros for each component with GOV.UK Frontend to make development easier and faster, but this also makes it harder for developers to manually alter the raw HTML. We'll need to add a new parameter to Nunjucks to help them out.

Reviewer: ```suggestion
However, this only works for developers who are writing raw HTML. We include Nunjucks macros for each component with GOV.UK Frontend to make development easier and faster, but this also makes it harder for developers to manually alter the raw HTML. We'll add a new parameter to Nunjucks to help them out.
```

## Adding a Nunjucks parameter

> Although most components support adding arbitrary attributes and values via the `attributes` parameter, not all of them do. This method is also more verbose compared to having a dedicated Nunjucks parameter.

Reviewer: ```suggestion
Most, but not all, components support adding arbitrary attributes and values through the `attributes` parameter. This method is also more verbose compared to having a dedicated Nunjucks parameter.
```

Let's update our previous bit of HTML to make the `data-*` attribute changeable.

```nunjucks
<div
  class="govuk-accordion"
  data-module="govuk-accordion"
  {%- if params.rememberExpanded %} data-remember-expanded="{{ params.rememberExpanded | escape }}"{% endif %}>
  ...
</div>
```

> The above checks for the existence of the `rememberExpanded` parameter and adds the `data-remember-expanded` attribute if it's present. It then passes in the developer's defined value, running it through [Nunjucks' native `escape` filter](https://mozilla.github.io/nunjucks/templating.html#escape-aliased-as-e) to ensure that values can't break our HTML.

Reviewer: ```suggestion
The above code checks for the existence of the `rememberExpanded` parameter. If the parameter is present it adds the `data-remember-expanded` attribute. It then passes in the developer's defined value, running it through [Nunjucks' native `escape` filter](https://mozilla.github.io/nunjucks/templating.html#escape-aliased-as-e) to make sure that values can't break our HTML.
```

We can now call the Accordion's Nunjucks macro with our new `rememberExpanded` parameter:

```nunjucks
{% from "govuk/components/accordion/macro.njk" import govukAccordion %}

{{ govukAccordion({
  id: "accordion",
  rememberExpanded: false,
  items: [...]
}) }}
```

> And that's pretty much it!

Reviewer: Although I want this line to stay because it's cheery, I think there needs to be a 'wrap up' line. Something about the benefit of using this method perhaps?

## Naming configuration options

### In JavaScript

The option in JavaScript should be written in camelCase.

If the option is a boolean option that accepts `true` or `false` as values, favour affirmative language when naming the option. This avoids the use of double-negatives being used to enable something. For example, `rememberExpanded: true` is easier to understand than `forgetExpanded: false`.

### In HTML

> `data-*` attributes in HTML should use the same name as JavaScript, converted into kebab-case. This is because the `dataset` property in JavaScript automatically converts between these two formats for us.

Reviewer: ```suggestion
`data-*` attributes in HTML should use the same name as JavaScript, converted into kebab-case. This is because the `dataset` property in JavaScript automatically converts between these two formats.
```

> > **Warning**: This automatic conversion doesn't happen for numbers, punctuation, or other non-Latin alphabet characters. For this reason, avoid using numbers, or spell them out as words instead, when naming configuration options.

Reviewer: ```suggestion
> **Warning**: This automatic conversion does not happen for numbers, punctuation, or other non-Latin alphabet characters. For this reason, when you're naming configuration options, avoid using numbers or spell them out as words.
```

In our example, `rememberExpanded` becomes `data-remember-expanded`.

### In Nunjucks

> Unlike the `data-*` attribute in HTML and our use of `dataset` in JavaScript, there is no intrinsic link between the Nunjucks parameter name and the names used elsewhere. The Nunjucks parameter can therefore be named different, if convenient.

Author: In our Nunjucks code and documentation, we encourage the names of prosaic Nunjucks parameters to include either 'text' or 'html' in their name to indicate how the value will be treated ('html' ones will go through an HTML parser, 'text' ones will not). 

When this new configuration API was introduced, we continued doing this for Nunjucks parameter naming. 

So for example, the Accordion uses `i18n.hideAllSections` in JavaScript, `data-i18n.hide-all-sections` in HTML, but then changes to `hideAllSectionsText` in Nunjucks.

> Unlike the `data-*` attribute in HTML and our use of `dataset` in JavaScript, there is no intrinsic link between the Nunjucks parameter name and the names used elsewhere. The Nunjucks parameter can therefore be named different, if convenient.

Reviewer: Thank you!

> A common case is specifying whether a parameter accepts HTML or only plain text. For example, if a configuration option's value is inserted into the page using `innerText`, you might want to name the Nunjucks parameter something like `sectionLabelText` to indicate that HTML won't be rendered.

Reviewer: ```suggestion
A common case is specifying whether a parameter accepts HTML or only plain text. For example, if a configuration option's value is inserted into the page using `innerText`, you might want to name the Nunjucks parameter something like `sectionLabelText` to indicate that HTML will not be rendered.
```

There is more guidance on naming Nunjucks parameters in [the Nunjucks API documentation](https://github.com/alphagov/govuk-frontend/blob/main/docs/contributing/coding-standards/nunjucks-api.md#naming-options).

## Namespacing configuration options

> Configuration options in JavaScript and HTML can be grouped together by using namespaces; period-separated strings that prefix the configuration option name. Namespaces follow the same formats as other option names, being camelCase in JavaScript and kebab-case in HTML.

Reviewer: ```suggestion
You can group configuration options in JavaScript and HTML together by using namespaces; period-separated strings that prefix the configuration option name. Namespaces follow the same formats as other option names, being camelCase in JavaScript and kebab-case in HTML.
```
